src/type/product.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2019
+//     by  Centrality Investments Ltd.
+//     and Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::tm_std::*;
+
+use crate::{
+	form::{CompactForm, Form, MetaForm},
+	IntoCompact, NamedField, Registry, UnnamedField,
+};
+use derive_more::From;
+use serde::Serialize;
+
+/// A Product type (consisting of fields) e.g. a struct
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug, Serialize, From)]
+#[serde(bound = "F::Type: Serialize")]
+#[serde(rename_all = "lowercase")]
+pub enum TypeProduct<F: Form = MetaForm> {
+	/// A struct with named fields
+	Struct(TypeProductStruct<F>),
+	/// A tuple struct with unnamed fields
+	TupleStruct(TypeProductTupleStruct<F>),
+}
```

Comment:
Should be merged together with optional named fields. A builder checks that either none or all names are set. This way we have a more unified data structure and serialization.